lib/ansible/plugins/connection/vmware_tools.py

```diff
@@ -0,0 +1,486 @@
+        made to esx-svr-1.domain1.com, and the file is available for download from http://esx-svr-1.domain1.com/guestFile?id=1&token=1234, the URL returned may
+        be http://*/guestFile?id=1&token=1234. The client replaces the asterisk with the server name on which it invoked the call.
+
+        https://code.vmware.com/apis/358/vsphere#/doc/vim.vm.guest.FileManager.FileTransferInformation.html
+        """
+        return url.replace("*", self.vmware_host)
+
+    def _fetch_file_from_vm(self, guestFilePath):
+        try:
+            fileTransferInformation = self.fileManager.InitiateFileTransferFromGuest(vm=self.vm, auth=self.vm_auth, guestFilePath=guestFilePath)
+        except vim.fault.NoPermission as e:
+            raise AnsibleError("No Permission Error: %s %s" % (to_native(e.msg), to_native(e.privilegeId)))
+
+        url = self._fix_url_for_hosts(fileTransferInformation.url)
+        response = requests.get(url, verify=self.validate_certs, stream=True)
+
+        if response.status_code != 200:
+            raise AnsibleError("Failed to fetch file")
+
+        return response
+
+    def delete_file_in_guest(self, filePath):
+        """Delete file from VM."""
+        try:
+            self.fileManager.DeleteFileInGuest(vm=self.vm, auth=self.vm_auth, filePath=filePath)
+        except vim.fault.NoPermission as e:
+            raise AnsibleError("No Permission Error: %s %s" % (to_native(e.msg), to_native(e.privilegeId)))
+
+    def exec_command(self, cmd, in_data=None, sudoable=True):
+        """Execute command."""
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        stdout = self.create_temporary_file_in_guest(suffix=".stdout")
+        stderr = self.create_temporary_file_in_guest(suffix=".stderr")
+
+        guest_program_spec = self._get_guest_program_spec(cmd, stdout, stderr)
+
+        try:
+            pid = self.processManager.StartProgramInGuest(vm=self.vm, auth=self.vm_auth, spec=guest_program_spec)
+        except vim.fault.NoPermission as e:
+            raise AnsibleError("No Permission Error: %s %s" % (to_native(e.msg), to_native(e.privilegeId)))
+        except vim.fault.FileNotFound as e:
+            raise AnsibleError("StartProgramInGuest Error: %s" % to_native(e.msg))
+
+        pid_info = self._get_pid_info(pid)
+
+        while pid_info.endTime is None:
+            sleep(self.get_option("exec_command_sleep_interval"))
+            pid_info = self._get_pid_info(pid)
+
+        stdout_response = self._fetch_file_from_vm(stdout)
+        self.delete_file_in_guest(stdout)
+
+        stderr_response = self._fetch_file_from_vm(stderr)
+        self.delete_file_in_guest(stderr)
+
+        return pid_info.exitCode, stdout_response.text, stderr_response.text
+
+    def fetch_file(self, in_path, out_path):
+        """Fetch file."""
+        super(Connection, self).fetch_file(in_path, out_path)
+
+        makedirs_safe(dirname(out_path))
```

Comment:
@jamescassell I think we should remove this as makedirs_safe is not safe. We should not create any directory as user should ensure creation of directory rather than creating them silently.

Reply:
s/user/caller/  (although user creation really is the only safe method if the file is in a world-writable directory).

Reply:
Yup. Totally agreed.

Reply:
Done.